bot/cogs/eval.py

```diff
@@ -12,66 +19,153 @@
+
+        self.stdout.seek(0)
+        text = self.stdout.read()
+        self.stdout.close()
+        self.stdout = io.StringIO()
+
+        if text:
+            res += text + "\n"
+
+        if out is None:
+            # No output, return the input statement
+            return (res, None)
+
+        res += f"Out[{self.ln}]: "
+
+        if isinstance(out, discord.Embed):
+            # We made an embed? Send that as embed
+            res += "<Embed>"
+            res = (res, out)
 
-        try:
-            rvalue = await self.interpreter.run(code, ctx, io)
-        except Exception as e:
-            await ctx.send(
-                f"{ctx.author.mention} **Code**\n"
-                f"```py\n{code}```\n\n"
-                f"**Error**\n```{e}```"
-            )
         else:
-            out_message = (
-                f"{ctx.author.mention} **Code**\n"
-                f"```py\n{code}\n```"
-            )
-
-            output = io.getvalue()
-
-            if output:
-                out_message = (
-                    f"{out_message}\n\n"
-                    f"**Output**\n```{output}```"
-                )
-
-            if rvalue is not None:
-                out_message = (
-                    f"{out_message}\n\n"
-                    f"**Returned**\n```py\n{repr(rvalue)}\n```"
-                )
-
-            await ctx.send(out_message)
+            if (isinstance(out, str) and
+                    out.startswith("Traceback (most recent call last):\n")):
+                # Leave out the traceback message
+                out = "\n" + "\n".join(out.split("\n")[1:])
+
+            pretty = (pprint.pformat(out, compact=True, width=60)
+                      if not isinstance(out, str) else str(out))
+
+            if pretty != str(out):
+                # We're using the pretty version, start on the next line
+                res += "\n"
+
+            if pretty.count("\n") > 20:
+                # Text too long, shorten
+                li = pretty.split("\n")
+                pretty = "\n".join(li[:3]) + "\n ...\n" + "\n".join(li[-3:])
+
+            # Add the output
+            res += pretty
+            res = (res, None)
+
+        return res
+
+    async def _eval(self, ctx, code):
+        self.ln += 1
+
+        if code.startswith("exit"):
+            self.ln = 0
+            self.env = {}
+            return await ctx.send(f"```Reset history!```")
+
+        env = {
+            "message": ctx.message,
+            "author": ctx.message.author,
+            "channel": ctx.channel,
+            "guild": ctx.guild,
+            "ctx": ctx,
+            "self": self,
+            "bot": self.bot,
+            "inspect": inspect,
+            "discord": discord,
+            "contextlib": contextlib
+        }
+
+        self.env.update(env)
+
+        # Ignore this shitcode, it works
+        _code = """
+async def func():
+    try:
+        with contextlib.redirect_stdout(self.stdout):
```

Comment:
Will do this mess up printing stuff to stdout from the rest of the bot while the eval is being processed?

Reply:
Contextlib *should* handle that. I've personally never had such issue.